subsys/shell/backends/shell_mqtt.c

```diff
@@ -0,0 +1,824 @@
+/*
+ * Copyright (c) 2022 G-Technologies Sdn. Bhd.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <shell/shell_mqtt.h>
+#include <init.h>
+#include <logging/log.h>
+#include <string.h>
+#include <stdio.h>
+#include <drivers/hwinfo.h>
+
+SHELL_MQTT_DEFINE(shell_transport_mqtt);
+SHELL_DEFINE(shell_mqtt, "", &shell_transport_mqtt,
+	     CONFIG_SHELL_BACKEND_MQTT_LOG_MESSAGE_QUEUE_SIZE,
+	     CONFIG_SHELL_BACKEND_MQTT_LOG_MESSAGE_QUEUE_TIMEOUT, SHELL_FLAG_OLF_CRLF);
+
+LOG_MODULE_REGISTER(shell_mqtt, CONFIG_SHELL_MQTT_LOG_LEVEL);
+
+#define NET_EVENT_MASK (NET_EVENT_L4_CONNECTED | NET_EVENT_L4_DISCONNECTED)
+#define CONNECT_TIMEOUT_MS 2000
+#define LISTEN_TIMEOUT_MS 500
+#define MQTT_SEND_DELAY_MS K_MSEC(100)
+#define PROCESS_INTERVAL K_SECONDS(2)
+#define SHELL_MQTT_WORKQ_STACK_SIZE 2048
+
+#ifdef CONFIG_SHELL_MQTT_SERVER_USERNAME
+#define MQTT_USERNAME CONFIG_SHELL_MQTT_SERVER_USERNAME
+#else
+#define MQTT_USERNAME NULL
+#endif /* CONFIG_SHELL_MQTT_SERVER_USERNAME */
+
+#ifdef CONFIG_SHELL_MQTT_SERVER_PASSWORD
+#define MQTT_PASSWORD CONFIG_SHELL_MQTT_SERVER_PASSWORD
+#else
+#define MQTT_PASSWORD NULL
+#endif /*SHELL_MQTT_SERVER_PASSWORD */
+
+struct shell_mqtt *sh_mqtt;
+K_KERNEL_STACK_DEFINE(sh_mqtt_workq_stack, SHELL_MQTT_WORKQ_STACK_SIZE);
+
+static void mqtt_evt_handler(struct mqtt_client *const client, const struct mqtt_evt *evt);
+
+static inline int sh_mqtt_work_reschedule(struct k_work_delayable *dwork, k_timeout_t delay)
+{
+	return k_work_reschedule_for_queue(&sh_mqtt->workq, dwork, delay);
+}
+
+static inline int sh_mqtt_work_submit(struct k_work *work)
+{
+	return k_work_submit_to_queue(&sh_mqtt->workq, work);
+}
+
+/* Lock the context of the shell mqtt */
+static inline int sh_mqtt_context_lock(k_timeout_t timeout)
+{
+	return k_mutex_lock(&sh_mqtt->lock, timeout);
+}
+
+/* Unlock the context of the shell mqtt */
+static inline void sh_mqtt_context_unlock(void)
+{
+	k_mutex_unlock(&sh_mqtt->lock);
+}
+
+bool __weak shell_mqtt_get_devid(char *id, int id_max_len)
+{
+	uint8_t hwinfo_id[DEVICE_ID_BIN_MAX_SIZE];
+	ssize_t length;
+
+	length = hwinfo_get_device_id(hwinfo_id, DEVICE_ID_BIN_MAX_SIZE);
+	if (length <= 0) {
+		return false;
+	}
+
+	(void)memset(id, 0, id_max_len);
+	length = bin2hex(hwinfo_id, (size_t)length, id, id_max_len - 1);
+
+	return length > 0;
+}
+
+static void prepare_fds(void)
+{
+	if (sh_mqtt->mqtt_cli.transport.type == MQTT_TRANSPORT_NON_SECURE) {
+		sh_mqtt->fds[0].fd = sh_mqtt->mqtt_cli.transport.tcp.sock;
+	}
+
+	sh_mqtt->fds[0].events = ZSOCK_POLLIN;
+	sh_mqtt->nfds = 1;
+}
+
+static void clear_fds(void)
+{
+	sh_mqtt->nfds = 0;
+}
+
+static int wait(int timeout)
+{
+	int rc = 0;
+
+	if (sh_mqtt->nfds > 0) {
+		rc = zsock_poll(sh_mqtt->fds, sh_mqtt->nfds, timeout);
+		if (rc < 0) {
+			LOG_ERR("poll error: %d", errno);
+		}
+	}
+
+	return rc;
+}
+
+/* Query IP address for the broker URL */
+static int get_mqtt_broker_addrinfo(void)
+{
+	int rc;
+
+	sh_mqtt->hints.ai_family = AF_INET;
```

Comment:
Couldn't `hints` structure be just declared on a stack, since it's just temporarily used for `getaddrinfo()` call? I think there's no need to bloat the shell context structure.

Reply:
> Couldn't hints structure be just declared on a stack, since it's just temporarily used for `getaddrinfo()` call?

I think you are correct, I don't actually know which arg need to persist and which doesn't, by 'stack', do you mean I can do it like so?

```c
/* Query IP address for the broker URL */
static int get_mqtt_broker_addrinfo(void)
{
	int rc;
	struct zsock_addrinfo hints;

	hints.ai_family = AF_INET;
	hints.ai_socktype = SOCK_STREAM;
	hints.ai_protocol = 0;

	if (sh_mqtt->haddr != NULL) {
		zsock_freeaddrinfo(sh_mqtt->haddr);
	}

	rc = zsock_getaddrinfo(CONFIG_SHELL_MQTT_SERVER_ADDR,
			       STRINGIFY(CONFIG_SHELL_MQTT_SERVER_PORT), &hints,
			       &sh_mqtt->haddr);
	if (rc == 0) {
		LOG_INF("DNS%s resolved for %s:%d", "", CONFIG_SHELL_MQTT_SERVER_ADDR,
			CONFIG_SHELL_MQTT_SERVER_PORT);

		return 0;
	}

	LOG_ERR("DNS%s resolved for %s:%d, retrying", " not", CONFIG_SHELL_MQTT_SERVER_ADDR,
		CONFIG_SHELL_MQTT_SERVER_PORT);

	return rc;
}
```

Reply:
fixed

Reply:
Sorry for late reply, yes by stack I meant to declara variable as you suggested. Something seems wrong though with the current implementation, it seems that you still use the struct member for hints, but it has been removed?

Reply:
I got confused with my local branch https://github.com/ycsin/zephyr/commit/3f26274db17120fa93d288e731e2e24ba4e37025#diff-a89e0b5c4a3caf077920c63b73e44c8aa46dc7f5ea840f3bd9c67613212c0e2aR127

Will sync them soon

Reply:
synced